### PR TITLE
[REST API] Magic Link for WordPress.com authentication - part 2

### LINKED ISSUE #8427: [REST API] Magic login flow for Jetpack connection

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #8427 

### Description
This PR is the second part for handling Magic Link authentication for WordPress.com connection, and it has the following the changes:

1. Simplification of Magic Link intercepting screen:
 Instead of having an architecture that uses the flow `MagicLinkInterceptActivity` -> `LoginActivity` -> `MagicLinkInterceptFragment` -> `MainActivity`, I simplified to `MagicLinkInterceptActivity` -> `MainActivity`.
 This part was handled in the three first commits: 94287cb047bdb994a7a48bca5574170ad38de469, 2bc19c8f8968c7717169442ec0d727a6303ca1c5.
2. A minor refactoring of `MagicLinkInterceptViewModel` to use events instead of multiple LiveDatas, this is now more consistent with the rest of the code base ce9880d0cb9dd1bf0dbb5403b947473ad959f6d2.
There is still room for improvement, but I opted for a simple change, for now, to move forward with the project.
2. The navigation from `MagicLinkInterceptActivity` to continue the Jetpack activation flow.

I suggest reviewing the three first commits about the refactoring separately.

**Note**: Jetpack installation is not still adapted for this flow, so it's expected that it fails.

##### Jetpack Not Installed
1. Sign in to the app using a site that doesn't have Jetpack.
2. Click on the Jetpack benefits banner, then click on "Login to continue"
3. Enter your wordpress.com email.
4. If the account is not passwordless, click on "or continue using Magic Link"
5. On your device, click on the Magic Link you received in your email (tip: you can also copy it and open it in the phone's browser like the below videos).
6. Confirm that the Jetpack installation starts.

##### Jetpack Not Connected
1. Sign in to the app using a site that has Jetpack but it's not connected.
2. Click on the Jetpack benefits banner, then click on "Login to continue"
3. Enter your wordpress.com email.
4. If the account is not passwordless, click on "or continue using Magic Link"
5. On your device, click on the Magic Link you received in your email (tip: you can also copy it and open it in the phone's browser like the below videos).
6. Confirm that the Jetpack connection starts.

##### Jetpack Connected
1. Sign in to the app using a non-Jetpack site (either not installed or not connected).
2. In a browser, install Jetpack then connect it.
3. Go back to the app.
4. Click on the Jetpack benefits banner, then click on "Login to continue" 
4. In the password screen, click on "or continue using Magic Link"
5. On your device, click on the Magic Link you received in your email (tip: you can also copy it and open it in the phone's browser like the below videos).
6. Confirm that the Jetpack connection starts.

##### Jetpack Not Installed
https://user-images.githubusercontent.com/1657201/222708322-fe4e5a5e-7503-43ab-bf9b-c97246d463f5.mov

##### Jetpack Not Connected
https://user-images.githubusercontent.com/1657201/222708393-666f9a7f-3b39-45f9-8806-4621f8aa449f.mov

##### Jetpack Connected
https://user-images.githubusercontent.com/1657201/222708336-37db9e33-ea4a-41e1-acf3-d0d6615750d2.mov

##### Error scenario
https://user-images.githubusercontent.com/1657201/222711189-48aa08c4-cbbe-4a3a-80d7-7124472060e1.mov

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComPostLoginViewModel.kt [hunk 1]
```diff
@@ -15,25 +15,30 @@ open class JetpackActivationWPComPostLoginViewModel(
...
-                    onFailure = {
+            // Attempt returning the site from the DB if it exists, otherwise fetch it from API
+            val jetpackSite = jetpackActivationRepository.getJetpackSiteByUrl(siteUrl)
+                .takeIf { it?.hasWooCommerce == true }
+                ?: jetpackActivationRepository.fetchJetpackSite(siteUrl)
+                    .getOrElse {
                         triggerEvent(ShowSnackbar(R.string.error_generic))
+                        return Result.failure(it)
                     }
-                )
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/jetpack/wpcom/JetpackActivationWPComPostLoginViewModel.kt:29]: The MagicLinkInterceptActivity doesn't continue until it fetches all WordPress.com sites, so at this step, normally we will have the site already available.
The error  handling is just for the unlikely scenario where the site doesn't appear in the `/me/sites` after Jetpack connection, this should never happen, but I wanted to offer a simple error handling for it.

For an example of this flow, check the video in the PR's description, to simulate it, I'm launching a deeplink manually that uses `source=wpcom-authentication`, but with a site that doesn't have Jetpack, so the site is not present after the fetch.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptActivity.kt [hunk 2]
```diff
@@ -1,27 +1,178 @@
...
 
     override fun onCreate(savedInstanceState: Bundle?) {
-        AndroidInjection.inject(this)
         super.onCreate(savedInstanceState)
 
         loginAnalyticsListener.trackLoginMagicLinkOpened()
 
+        setContentView(R.layout.login_magic_link_sent_screen)
+        retryButton = findViewById(R.id.login_open_email_client)
+        retryContainer = findViewById(R.id.login_magic_link_container)
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptActivity.kt:59]: Np, how about using view binding and removing the class fields?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptViewModel.kt [hunk 3]
```diff
@@ -71,4 +96,26 @@ class MagicLinkInterceptViewModel @Inject constructor(
         super.onCleared()
         magicLinkInterceptRepository.onCleanup()
     }
+
+    private fun MagicLinkSource.inferJetpackStatus(): JetpackStatus {
+        val isJetpackInstalled = this == MagicLinkSource.JetpackConnection ||
+            this == MagicLinkSource.WPComAuthentication
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptViewModel.kt]: np could be more readable `val isJetpackInstalled = this != MagicLinkSource.JetpackInstallation` ?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptActivity.kt [hunk 4]
```diff
@@ -1,27 +1,178 @@
...
+                ).toBundle()
+            )
+            .addDestination(
+                R.id.jetpackActivationMagicLinkHandlerFragment,
+                JetpackActivationMagicLinkHandlerFragmentArgs(
+                    jetpackStatus = event.jetpackStatus
+                ).toBundle()
+            )
+            .createPendingIntent()
+            .send()
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptActivity.kt:174]: Nice! TIL about building the stack with multiple destinations this way!

**Review by JorgeMucientes — APPROVED** [review]

**Comment by JorgeMucientes** [comment]
Great job @hichamboushaba 🏅. All scenarios work as expected. 
Thanks for extracting the logic for handling the deep link out of main activity. 

Left couple minor np, that you can perfectly ignore.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptViewModel.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptViewModel.kt]: Good point, done in [de3bcac](https://github.com/woocommerce/woocommerce-android/pull/8470/commits/de3bcac6b3f951157e87741157b247bc07cf0173)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptActivity.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/MagicLinkInterceptActivity.kt:59]: This screen needs more refactorings, if I can find time I'll try to clean it up a bit in the future, but not now 🙂.

**Pull request merged**